File: frontend/lib/models/teacher_assignment.dart
/// Teacher Assignment Model
/// Represents the assignment of a teacher to a course
class TeacherAssignment {
  final String id;
  final String teacherId;
  final String courseId;
  final String assignedBy;
  final DateTime assignedAt;
  final bool isActive;
  final String? notes;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  TeacherAssignment({
    required this.id,
    required this.teacherId,
    required this.courseId,
    required this.assignedBy,
    required this.assignedAt,
    this.isActive = true,
    this.notes,
    this.createdAt,
    this.updatedAt,
  });

  factory TeacherAssignment.fromJson(Map<String, dynamic> json) {
    return TeacherAssignment(
      id: json['_id'] ?? json['id'] ?? '',
      teacherId: json['teacherId'] ?? '',
      courseId: json['courseId'] ?? '',
      assignedBy: json['assignedBy'] ?? '',
      assignedAt: json['assignedAt'] != null
          ? DateTime.parse(json['assignedAt'])
          : DateTime.now(),
      isActive: json['isActive'] ?? true,
      notes: json['notes'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'teacherId': teacherId,
      'courseId': courseId,
      'assignedBy': assignedBy,
      'assignedAt': assignedAt.toIso8601String(),
      'isActive': isActive,
      'notes': notes,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }
}

File: backend/src/models/TeacherAssignment.js
const mongoose = require('mongoose');

const teacherAssignmentSchema = new mongoose.Schema({
  teacherId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'Teacher ID is required']
  },
  courseId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Course',
    required: [true, 'Course ID is required']
  },
  assignedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'Assigner ID is required']
  },
  assignedAt: {
    type: Date,
    default: Date.now
  },
  isActive: {
    type: Boolean,
    default: true
  },
  notes: {
    type: String,
    maxlength: [500, 'Notes cannot exceed 500 characters']
  }
}, {
  timestamps: true
});

// Ensure a teacher can only be assigned once to a course
teacherAssignmentSchema.index({ teacherId: 1, courseId: 1 }, { unique: true });
teacherAssignmentSchema.index({ teacherId: 1, isActive: 1 });
teacherAssignmentSchema.index({ courseId: 1, isActive: 1 });

module.exports = mongoose.model('TeacherAssignment', teacherAssignmentSchema);
